### PR TITLE
Week 4 hieu le

### PR DESCRIPTION
**Pull request by Hieuhuule**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [X ] JavaScript Coding Challenges
- [X ] Ruby Coding Challenges
- [X ] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: ruby-challenges.rb [hunk 1]
```diff
@@ -32,3 +54,15 @@
...
+# create a method that takes a string
+# use a conditional to determine if string is a palindrome
+# use .lowercase to make all the strings the same for conditional evaluation purposes
+# use string interpolation to print the string '#{string} is a palindrome'
+
+def palindrome str
+    new_str = str.downcase
+    rev_str = new_str.reverse
+    new_str == rev_str ? (p "#{str} is a palindrome") : (p "#{str} is not a palindrome")
+end
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:68]: good job on the ternary operator. 
Make sure you are performing a method call for each test variable

File: ruby-challenges.rb [hunk 2]
```diff
@@ -22,6 +35,15 @@
...
 
+# PSEUDO CODE
+# create a method that takes a string
+# iterate through the string
+# if charater is a vowel, delete it
+# it should implicitly return the new string
+
+def no_vowels string
+    p string.delete "aeiouAEIOU"
+end
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:46]: nice one-liner. 
Where is the method call?

File: ruby-challenges.rb [hunk 3]
```diff
@@ -10,8 +11,20 @@
...
 
 
+def even_or_odd num
+    num.even? ?  (p "#{num} is even") : (p "#{num} is odd")
+end
+
+even_or_odd num1
+even_or_odd num2
+even_or_odd num3
+
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:27]: ternary operators are awesome

File: interview-questions.md [hunk 4]
```diff
@@ -8,42 +8,46 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+Researched answer: I was mistaken, there's actually another portion that is called a block. That's the code between curly braces. If a code is between a do and end / or {}, then it's a block.
 
-Researched answer:
 
 5. What is an instance variable in Ruby?
 
-Your answer:
+Your answer: It's a variable that was made from a Class, while it has it's own unique hash identifier, it also has all the properties from it's Class, unless explicitly changed.
+
+Researched answer: An instance variable is declared by prefacing it with the @ symbol. Any instance of the class will be able to have scope accessed to the instanced variable.
```
Comment (SunkissedQueen) [inline interview-questions.md:40]: nice. More content can be your opinion with using instance variables, how to add default values, how to make it a required argument, etc. The examples help show that you have used the tool.

File: interview-questions.md [hunk 5]
```diff
@@ -8,42 +8,46 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Researched answer:
+Researched answer: Because Ruby uses implicit returns (returning the value of the last line automatically), people enjoy writting short, elegant code. Explicitly using 'return' can even lead to errors in Ruby!
 
 4. What is a block in Ruby?
 
-Your answer:
+Your answer: it's everything between a 'def' and 'end', it's the code, the script, and the instructions for the interpreter.
+
+Researched answer: I was mistaken, there's actually another portion that is called a block. That's the code between curly braces. If a code is between a do and end / or {}, then it's a block.
```
Comment (SunkissedQueen) [inline interview-questions.md:33]: yes. do you have any preference between the two?

File: interview-questions.md [hunk 6]
```diff
@@ -8,42 +8,46 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Researched answer:
 
 3. Ruby has an implicit return. What does that mean?
 
-Your answer:
+Your answer: In JavaScript you must explicitly state return in order for scripts to return anything, however, in Ruby, every line has a return, even if it's not displayed to the user.
 
-Researched answer:
+Researched answer: Because Ruby uses implicit returns (returning the value of the last line automatically), people enjoy writting short, elegant code. Explicitly using 'return' can even lead to errors in Ruby!
```
Comment (SunkissedQueen) [inline interview-questions.md:28]: more content could be discussing those errors or giving an example of implicit return

File: interview-questions.md [hunk 7]
```diff
@@ -8,42 +8,46 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+Researched answer: OOP programming is more abstract because it has a focus on operating on keys directly instead of functional programming's declarative approach that requires explicit functions. As such, they have different methods for storing and handling data.
 
-Researched answer:
 
 2. What is the difference between a Float and an Integer in Ruby?
 
-Your answer:
+Your answer: A float is a decimal value and an Integer is a whole number value.
+
+Researched answer: It's a simple question and the research answer confirms that any number written without decimals is an integer and if it has a decimal, it counts as a float. So here's a joke: Why should you never argue with a decimal? It's because they always have a point.
```
Comment (SunkissedQueen) [inline interview-questions.md:20]: like the humor...very appropriate for this question. Additional content could be some other facts about integers in Ruby.

File: interview-questions.md [hunk 8]
```diff
@@ -8,42 +8,46 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What is object-oriented programming? How is it different than functional programming?
 
-Your answer:
+Your answer: Object-oriented programming focuses on using key/value pairs whereas functional programming puts an emphasis on using higher order functions?
+
+Researched answer: OOP programming is more abstract because it has a focus on operating on keys directly instead of functional programming's declarative approach that requires explicit functions. As such, they have different methods for storing and handling data.
```
Comment (SunkissedQueen) [inline interview-questions.md:13]: nice. additional content could be sharing some websites that use those programming styles

File: code-challenges.test.js [hunk 9]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+        noDuplicates.push(value)
+       } 
+    })
+    arr2.map((value) => {
+        if (noDuplicates.includes(value) === false) {
+         noDuplicates.push(value)
+        } 
+     })
+    return noDuplicates
+}
```
Comment (SunkissedQueen) [inline code-challenges.test.js:139]: Unique approach. Did this function pass? If so, please refactor. Go back and see if there is way that you can practice the DRY principle.  May be there is a way to combine the two arrays and then return an array with the values that true to a condition that is set? You can do all of this without needing an empty array to store the values.

File: code-challenges.test.js [hunk 10]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+const voteCounter = (obj) => {
+    let voteSum = (obj.upVotes - obj.downVotes)
+    return voteSum};
+
 
 // --------------------3) Create a function that takes in two arrays as arguments and returns one array with no duplicate values. STRETCH: Use the spread operator to pass in a dynamic number of arguments.
 
+// PSEUDO CODE
+// create a function that takes in two arrays
+// iterate over each array
```
Comment (SunkissedQueen) [inline code-challenges.test.js:101]: What element are you looking at during the iteration

File: code-challenges.test.js [hunk 11]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+// if value doesn't exist in third array, add to third array
+// push the value into a third array
+// return the array with NO DUPLICATE VALUES
+// STRETCH: No thanks.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:105]: try to stretch...highly suggest you look at the spread operator and new set to see how a refactor would look using those tools

File: code-challenges.test.js [hunk 12]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+    }
+    shuffle(rm_First_Element)
+}
+
 
 // --------------------2) Create a function that takes in an object that contains up votes and down votes and returns the net total of votes.
 
+// PSEUDO CODE
+// create a function that takes an object of key/value vote: up or down
+// return total votes
```
Comment (SunkissedQueen) [inline code-challenges.test.js:68]: suggest including how you will abstract the values and the type of calculation you will use to obtain the total

File: code-challenges.test.js [hunk 13]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
 const colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
 // Expected output example (can be a different order): ["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]
 
+    // PSUEDO CODE
+    // create function that takes an array
+    // removes first item from array
+    // shuffles array
```
Comment (SunkissedQueen) [inline code-challenges.test.js:26]: recommend explaining how you will shuffle the array

File: code-challenges.test.js [hunk 14]
```diff
@@ -21,30 +20,120 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+    // return new array
+    // for the test, expect statement should be flexible on order of returened array in expected output
+
+    describe('rm_First_Element_Shuffle_Array', () => {
+        it('takes an array, removes the first element and shuffles the remaining contents', () => {
+            expect(rm_First_Element_Shuffle_Array(colors1)).not.toEqual(expect.arrayContaining(["yellow", "blue", "pink", "green"]));
```
Comment (SunkissedQueen) [inline code-challenges.test.js:32]: When using the `not` jest matcher, you will need to only type the value that you are not expecting to show in the array. Did your test pass after creating the function?

**Review by SunkissedQueen — APPROVED** [review]
Hieu, you are approved! Thank you for giving this assessment an honest effort. You are a confident developer. I would like to see you pour that same pizazz that you did in the ruby challenge onto the jest section. You use ternary operators!!! Next week assessment requests:
- Ensure your pseudocode is discussing what element you are using in the iteration (value, index, array), what you plan to do with that element, how you plan to transform any data types, refactor
- On the interview answers, maintain your approach with defining the tool...let's also share a bit of your personal experience or opinion with using the tool

File: ruby-challenges.rb (hunk 1 above)
Comment (Hieuhuule) [reply ruby-challenges.rb:68]: Ah, I had tested it in a different document and once I got it to pass, I put the code in there. I thought the method call was extraneous.

File: ruby-challenges.rb (hunk 2 above)
Comment (Hieuhuule) [reply ruby-challenges.rb:46]: Ah, I had tested it in a different document and once I got it to pass, I put the code in there. I thought the method call was extraneous.

File: code-challenges.test.js (hunk 9 above)
Comment (Hieuhuule) [reply code-challenges.test.js:139]: Yes, the function passed the test.

File: code-challenges.test.js (hunk 10 above)
Comment (Hieuhuule) [reply code-challenges.test.js:101]: I was looking at the value during iteration and adding it to the formerly empty array if it didn't exist in there previously.

File: code-challenges.test.js (hunk 13 above)
Comment (Hieuhuule) [reply code-challenges.test.js:26]: This shuffle method was cowboy code

File: interview-questions.md (hunk 5 above)
Comment (Hieuhuule) [reply interview-questions.md:33]: I prefer the def and end, it's more clear

File: code-challenges.test.js (hunk 14 above)
Comment (Hieuhuule) [reply code-challenges.test.js:32]: yes, the test passed

**Pull request merged**